x/allocation/keeper/keeper.go

```diff
@@ -0,0 +1,100 @@
+package keeper
+
+import (
+	"fmt"
+
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	"github.com/noislabs/noisd/x/allocation/types"
+
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+type Keeper struct {
+	cdc      codec.BinaryCodec
+	storeKey sdk.StoreKey
+
+	accountKeeper types.AccountKeeper
+	bankKeeper    types.BankKeeper
+	stakingKeeper types.StakingKeeper
+
+	paramstore paramtypes.Subspace
+}
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey sdk.StoreKey,
+	accountKeeper types.AccountKeeper, bankKeeper types.BankKeeper, stakingKeeper types.StakingKeeper,
+	ps paramtypes.Subspace,
+) *Keeper {
+	// set KeyTable if it has not already been set
+	if !ps.HasKeyTable() {
+		ps = ps.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		accountKeeper: accountKeeper, bankKeeper: bankKeeper, stakingKeeper: stakingKeeper,
+		paramstore: ps,
+	}
+}
+
+// DistributeInflation distributes module-specific inflation
+func (k Keeper) DistributeInflation(ctx sdk.Context) error {
+	// Fee collector module account account contains newly minted coins and collected fees from transactions
+	blockInflationAddr := k.accountKeeper.GetModuleAccount(ctx, authtypes.FeeCollectorName).GetAddress()
+	blockInflation := k.bankKeeper.GetBalance(ctx, blockInflationAddr, k.stakingKeeper.BondDenom(ctx))
+
+	// if there is no inflation, return
+	if blockInflation.IsZero() {
+		return nil
+	}
+	params := k.GetParams(ctx)
+	proportions := params.DistributionProportions
+	// fund randomness rewards address
+	randomnessRewardsCoin := k.GetProportions(ctx, blockInflation, proportions.RandomnessRewards)
+	randomnessRewardsReceiver, err := sdk.AccAddressFromBech32(params.RandomnessRewardsReceiver)
+	if err != nil {
+		return err
+	}
+	if !randomnessRewardsCoin.IsZero() {
+		k.bankKeeper.SendCoinsFromModuleToAccount(ctx, authtypes.FeeCollectorName, randomnessRewardsReceiver, sdk.NewCoins(randomnessRewardsCoin))
+	}
+	// fund validator rewards pool
+	validatorRewardsCoins := k.GetProportions(ctx, blockInflation, proportions.ValidatorRewards)
+	if !validatorRewardsCoins.IsZero() {
+		k.bankKeeper.SendCoinsFromModuleToModule(ctx, authtypes.FeeCollectorName, types.ValidatorRewardsPool, sdk.NewCoins(validatorRewardsCoins))
+	}
+	return nil
+}
+
+func (k Keeper) DistributeValidatorRewards(ctx sdk.Context) error {
+	validators := k.stakingKeeper.GetLastValidators(ctx)
+	validValidators := make([]stakingtypes.Validator, 0)
+	for _, validator := range validators {
+		// skip jailed and not bonded
+		if validator.Jailed || validator.Status != stakingtypes.Bonded {
+			continue
+		}
+		// TODO: retrieve self-delegation
```

Comment:
What does this bonding status mean? I don't really understand it.

```
// BondStatus is the status of a validator.
enum BondStatus {
  option (gogoproto.goproto_enum_prefix) = false;

  // UNSPECIFIED defines an invalid validator status.
  BOND_STATUS_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "Unspecified"];
  // UNBONDED defines a validator that is not bonded.
  BOND_STATUS_UNBONDED = 1 [(gogoproto.enumvalue_customname) = "Unbonded"];
  // UNBONDING defines a validator that is unbonding.
  BOND_STATUS_UNBONDING = 2 [(gogoproto.enumvalue_customname) = "Unbonding"];
  // BONDED defines a validator that is bonded.
  BOND_STATUS_BONDED = 3 [(gogoproto.enumvalue_customname) = "Bonded"];
}
```

Maybe we don't need the self-delegation value if we have Bonded?

Reply:
moving to get all bonded validators for now